Source/items.cpp

```diff
@@ -1741,76 +1741,99 @@ void DrawUniqueInfoWindow(const Surface &out)
-			AddPanelString(_("Right-click to use"));
+		if (!invflag) {
+			AddPanelString(_("Open inventory to use"));
+
 		} else {
-			if (!invflag) {
-				AddPanelString(_("Open inventory to use"));
-			} else {
-				AddPanelString(_("Activate to use"));
-			}
+			AddPanelString(_("Activate to use"));
 		}
+	} else if (item._iMiscId == IMISC_SCROLL) {
+		AddPanelString(_("Select from spell book, then"));
+		AddPanelString(_("cast to read"));
+	} else {
+		AddPanelString(_("Activate to read"));
 	}
-	if (IsAnyOf(item._iMiscId, IMISC_BOOK, IMISC_NOTE)) {
-		if (ControlMode == ControlTypes::KeyboardAndMouse) {
-			AddPanelString(_("Right-click to read"));
+}
+
+void printItemMiscGamepad(const Item &item, const bool isOil, const bool isCastOnTarget, const bool usingDualShock)
+{
+	if (item._iMiscId == IMISC_MAPOFDOOM) {
+		if (usingDualShock) {
+			AddPanelString(_("Triangle to view"));
 		} else {
-			AddPanelString(_("Activate to read"));
+			AddPanelString(_("Y to view"));
 		}
-	}
-	if (item._iMiscId == IMISC_MAPOFDOOM) {
-		if (ControlMode == ControlTypes::KeyboardAndMouse) {
-			AddPanelString(_("Right-click to view"));
+	} else if (isOil) {
+		PrintItemOil(item._iMiscId);
+		if (!invflag) {
+			AddPanelString(_("Open inventory to use"));
+
+		} else if (usingDualShock) {
+			AddPanelString(_("Triangle to use"));
 		} else {
-			AddPanelString(_("Activate to view"));
+			AddPanelString(_("Y to use"));
 		}
+	} else if (isCastOnTarget) {
+		AddPanelString(_("Select from spell book, then"));
+		if (usingDualShock)
+			AddPanelString(_("Square to read"));
+		else
+			AddPanelString(_("X to read"));
+	} else if (IsAnyOf(item._iMiscId, IMISC_BOOK, IMISC_NOTE, IMISC_SCROLL)) {
+		if (usingDualShock)
+			AddPanelString(_("Triangle to read"));
+		else
+			AddPanelString(_("Y to read"));
 	}
+}
+
+void PrintItemMisc(const Item &item)
+{
 	if (item._iMiscId == IMISC_EAR) {
 		AddPanelString(fmt::format(fmt::runtime(pgettext("player", "Level: {:d}")), item._ivalue));
+		return;
 	}
 	if (item._iMiscId == IMISC_AURIC) {
 		AddPanelString(_("Doubles gold capacity"));
+		return;
+	}
+	const bool isOil = (item._iMiscId >= IMISC_USEFIRST && item._iMiscId <= IMISC_USELAST)
+	    || (item._iMiscId > IMISC_OILFIRST && item._iMiscId < IMISC_OILLAST)
+	    || (item._iMiscId > IMISC_RUNEFIRST && item._iMiscId < IMISC_RUNELAST);
+	const bool isCastOnTarget = (item._iMiscId == IMISC_SCROLLT && item._iSpell != SPL_FLASH)
+	    || item._iSpell == SPL_TOWN
+	    || item._iSpell == SPL_IDENTIFY
+	    || TargetsMonster(item._iSpell);
+
+	if (ControlMode == ControlTypes::KeyboardAndMouse) {
+		printItemMiscKBM(item, isOil, isCastOnTarget);
+	} else if (ControlMode == ControlTypes::VirtualGamepad || GamepadType == SDL_CONTROLLER_TYPE_VIRTUAL) {
+		printItemMiscVirtualGamepad(item, isOil);
+	} else {
+		const bool usingDualShock = IsAnyOf(GamepadType, SDL_CONTROLLER_TYPE_PS3, SDL_CONTROLLER_TYPE_PS4, SDL_CONTROLLER_TYPE_PS5);

```

Comment:
Does a bool accurately represent the various types of layouts?

Is the brand name for a line of controllers introduced between 1997 and 2014 the best way to express a layout that was introduced by the original PlayStation in 1994 and still in use (and by other products)?

Reply:
Did a bit of research, and there are 3 main types:

- Nintendo (SNES 1990)
- PlayStation (1994)
- Xbox (2001)

The Nintendo layout goes back to NES 1985, but since it only had 8 of the 12 buttons we need it's not really relevant. Also they experimented a lot with the controller layout between 1995 and 2006, but we do not really support any of these yet.

Here are the enums that the SDL API provides grouped in to the ones that they belong to.

```
Nintendo:
    SDL_CONTROLLER_TYPE_NINTENDO_SWITCH_JOYCON_LEFT,
    SDL_CONTROLLER_TYPE_NINTENDO_SWITCH_JOYCON_RIGHT,
    SDL_CONTROLLER_TYPE_NINTENDO_SWITCH_JOYCON_PAIR,
    SDL_CONTROLLER_TYPE_NINTENDO_SWITCH_PRO,

PlayStation:
    SDL_CONTROLLER_TYPE_PS3,
    SDL_CONTROLLER_TYPE_PS4,
    SDL_CONTROLLER_TYPE_PS5,

Xbox:
    SDL_CONTROLLER_TYPE_GOOGLE_STADIA,
    SDL_CONTROLLER_TYPE_AMAZON_LUNA,
    SDL_CONTROLLER_TYPE_NVIDIA_SHIELD,
    SDL_CONTROLLER_TYPE_XBOX360,
    SDL_CONTROLLER_TYPE_XBOXONE,

Generic:
    SDL_CONTROLLER_TYPE_UNKNOWN = 0,
    SDL_CONTROLLER_TYPE_VIRTUAL,
```